[1단계 - 로또] 이지윤 미션 제출합니다.

dd-jiyun opened the pull request:
안녕하세요 주노 🙌

지난 자동차 미션에 이어 로또 미션도 화이팅 해보겠습니다! 잘 부탁드려요 🙇🏻‍♀️

---

## 🤔 고민한 부분

### 로또 번호 생성의 책임
로또 번호를 어떻게 생성할지에 대한 책임을 고민했습니다.
제가 정의한 Lotto는 스스로 번호를 생성하기보다는, 외부에서 주어진 번호를 검증하고 보유하는 객체로 생각했습니다.
그래서 번호 생성 책임은 LottoNumberGenerator 전략 객체에 위임했어요.

### 자동차 미션에서의 회고
자동차 미션에서는 Cars라는 일급 컬렉션이 Car들에게 한 번에 동일한 전략을 주입하는 방식이었어요.
이로 인해 각 Car마다 다른 전략을 주는 것이 어려웠던 경험이 있었습니다 😂

이번에도 Lottos가 생성 시점에 하나의 Generator만 주입받는 방식으로 구성되어 있다 보니 `만약 수동/자동 로또를 섞어서 생성하는 요구사항이 생기면 어떻게 할까?`라는 의문이 들어, 유연하게 대응할 수 있는 구조를 계속 고민 중입니다.

## 🧐 궁금한 부분
이번에는 실제 로또 구매 과정을 모델링하면서 객체로서의 책임을 부여하려고 노력했어요. 하지만 여전히 "객체의 역할"보다는 책임만을 기준으로 한 설계에 머물러 있는 건 아닌지 고민이 됩니다. 이 설계가 적절하게 객체 중심의 사고를 반영하고 있는지 궁금합니다!

dd-jiyun committed:
docs: 기능 구현 목록 작성

dd-jiyun committed:
feat: 로또 생성 및 검증 구현

dd-jiyun committed:
feat: 로또 번호를 감싸는 일급 컬렉션 추가 및 메서드 명 변경

dd-jiyun committed:
refactor: final 키워드 추가

dd-jiyun committed:
test: 로또 번호들 테스트 추가

dd-jiyun committed:
feat: 입력한 금액만큼 로또 번호 구매해주는 객체 생성

dd-jiyun committed:
feat: I/O 뷰 및 실행 메서드 추가

dd-jiyun committed:
test: 구매 금액 검증 테스트 추가

dd-jiyun committed:
refactor: 코드 컨벤션에 맞춰 수정

izzy80 commented:
꼼꼼하게 보려고 노력했습니다! 😄😄
미션 홧팅~~

지윤님 코드 보고 많이 배워갑니다 :)

izzy80 commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,21 @@
+        PurchaseAmount purchaseAmount = new PurchaseAmount(amount);
+
+        LottoNumberGenerator generator = new RandomNumberGenerator();
+        LottoStore store = new LottoStore(generator);
```
LottoStore라는 네이밍이 인상 깊었습니다!
저는 구매자의 입장에서만 생각했는데, 판매하는 쪽의 관점에서 책임을 분리한 게 설계적으로 더 명확하다는 느낌이 들었어요
덕분에 네이밍과 책임 분리에 대해 다시 한번 생각해보게 되었습니다! 좋은 아이디어 감사합니다 😊

izzy80 commented on src/main/java/domain/PurchaseAmount.java:
```diff
@@ -0,0 +1,34 @@
+
+public class PurchaseAmount {
+
+    private static final int LOTTO_PRICE = 1_000;
```
LOTTO_PRICE를 1_000처럼 언더스코어로 표현해주셔서 가독성이 훨씬 좋아진 것 같아요!
숫자가 금액처럼 길어질 때는 이런 방식이 확실히 더 읽기 쉽고, 실수도 줄일 수 있을 것 같네요

dd-jiyun replied:
맞아요! 사실 계산하는 것에는 차이는 없는데 구분해주는게 가독성이 좋더라구요 👍🏻

Choi-JJunho commented on README.md:
```diff
@@ -0,0 +1,16 @@
+## 🕹️ 로또
```
꼼꼼한 readme 좋네요!

Choi-JJunho commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    private void validate(final List<Integer> numbers) {
+        validateLottoCount(numbers);
+        validateDuplicateLottoNumber(numbers);
+        validateLottoRange(numbers);
```
갯수, 중복, 숫자범위검증까지 꼼꼼하게 잘 작성해주셨네요! 👍

Choi-JJunho commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,53 @@
+        }
+    }
+
+    private void validateLottoRange(final List<Integer> numbers) {
+        for (int num : numbers) {
+            if (num < MIN_LOTTO_NUMBER || MAX_LOTTO_NUMBER < num) {
+                throw new IllegalArgumentException("로또 번호는 1 ~ 45 사이의 값이어야 합니다.");
+            }
+        }
```
로또번호를 래핑하면 2depth를 벗어날 수 있을것 같은데 어떻게생각하시나요?

dd-jiyun replied:
제가 주노의 리뷰를 제대로 이해한 게 맞는지 모르겠어요! 

로또를 구성하는 번호를 한 번 더 원시값으로 감싸는 것을 말씀하시는 건가요?

Choi-JJunho replied:
네! 맞습니다!

dd-jiyun replied:
로또 번호를 래핑할 생각은 못했는데! 
각 번호를 검증하는 역할은 Lotto보다는 로또 번호 자체에서 하는 게 더 좋은 것 같아요 👍🏻

Choi-JJunho commented on src/main/java/strategy/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.Collections;
+import java.util.List;
+
+public class RandomNumberGenerator implements LottoNumberGenerator {
```
제가 생각하기에는 RandomNumberGenerator가 더 추상화된 개념이라고 생각이 드는데 어떻게 생각하시나요?

랜덤한 숫자를 생성 -> 1~45 중 하나를 생성하는 로또규칙에 맞는 랜덤숫자 전략

dd-jiyun replied:
음, 주노가 말한 것처럼 "어떻게 만들까?"라는 관점에서 보면 RandomNumberGenerator가 더 추상적인 개념이라는 점에 저도 공감해요!

그런데 저는 이번 구조를 설계할 때,
- 무엇을 만들까 ? -> 로또 번호
- 어떻게 만들까 ? -> 랜덤하게!

이런 흐름으로 먼저 `무엇`을 명확히 정해놓고, 그 다음에 `어떻게`를 결정하는 구조가 더 자연스럽다고 느꼈어요.

이후의 로또 미션에서 수동과 자동으로 나뉘는 것을 고려했을 때 `이 방식이 조금 더 유연하지 않을까?` 생각했어요. 혹시 다른 관점이 있다면 더 들어보고 싶어요!

Choi-JJunho replied:
좋은 흐름이라고생각해요!
지윤님 의견에 설득되었어요 ㅎㅎ
말씀해주신내용을 듣고보니 좋은 구조라고 생각이듭니다 👍

Choi-JJunho commented on src/test/java/domain/LottoStoreTest.java:
```diff
@@ -0,0 +1,28 @@
+        // then
+        assertThat(lottos.getValues())
+                .hasSize(3)
+                .allSatisfy(lotto -> assertThat(lotto.getNumbers()).containsExactly(1, 2, 3, 4, 5, 6));
```
다양한 테스트메소드를 잘 활용해주셨네요! 👍👍

Choi-JJunho reviewed (commented):
안녕하세요 지윤님~! 리뷰가 늦어서 죄송해요 ㅠㅠ

객체의 책임에 대해 많이 고민하신부분이 보이는 코드였습니다! 고민하신 책임에 대한 부분은 잘 작성해주신 것 같아 현재로서 피드백 드릴부분이 없을정도에요 ㅎㅎㅎ👍👍

앞으로 요구사항이 추가되는 것을 기대해보죠!

고생하셨습니다

dd-jiyun committed:
refactor: 로또 번호 원시값 포장

- 기존에는 로또 번호들을 반복문으로 1부터 45까지 포함되었는 지 확인하였지만, 로또 번호를 원시값으로 포장하여 스스로 값을 검증할 수 있게 책임을 분리하였습니다.

dd-jiyun commented on src/test/java/domain/LottoStoreTest.java:
```diff
@@ -23,6 +23,13 @@ void buy_shouldReturnCorrectNumberOfLottos() {
+                                new LottoNumber(3),
+                                new LottoNumber(4),
+                                new LottoNumber(5),
+                                new LottoNumber(6))
```
원시 값을 감싸니, 기존에 List.of로 작성했을 때보다 가독성이 떨어지는 느낌이 들었어요.

처음에는 `정적 팩토리 메서드를 사용하면 new로 객체를 생성하는 것보다 나아지지 않을까?`라는 생각으로 아래와 같이 수정해봤어요!

```java
.containsExactly(LottoNumber.of(1),
                 LottoNumber.of(2),
                 LottoNumber.of(3),
                 LottoNumber.of(4),
                 LottoNumber.of(5),
                 LottoNumber.of(6)));
```

그런데... 🤔 가독성 측면에서는 큰 개선이 없는 것 같더라고요.

그래서 `어떻게 하면 명확하게 의미를 표현하면서도, List.of 처럼 한 줄로 간결하게 쓸 수 있을까?`를 고민하다가
테스트 헬퍼 클래스라는 키워드를 알게 되서 적용 시켜 보려합니다!

혹시 이 외에도 테스트 코드의 가독성을 더 높일 수 있는 방법이 있을까요?

dd-jiyun committed:
test: 중복된 코드를 테스트 헬퍼 클래스로 분리

dd-jiyun committed:
fix: 원시값으로 포장 후 정렬 문제 수정

- 로또 번호를 원시값으로 포장한 후 객체가 그대로 반환되어 원하는 결과가 나오지 않아 수정하였습니다.
- 로또 번호가 정렬되지 않고 출력되는 문제를 Comparable을 구현하여 수정하였습니다.
